File: requirements/python/library/algebra/tensor/accelerator_tensor_inverse.py
###################################################################################
##                                            __ _      _     _                  ##
##                                           / _(_)    | |   | |                 ##
##                __ _ _   _  ___  ___ _ __ | |_ _  ___| | __| |                 ##
##               / _` | | | |/ _ \/ _ \ '_ \|  _| |/ _ \ |/ _` |                 ##
##              | (_| | |_| |  __/  __/ | | | | | |  __/ | (_| |                 ##
##               \__, |\__,_|\___|\___|_| |_|_| |_|\___|_|\__,_|                 ##
##                  | |                                                          ##
##                  |_|                                                          ##
##                                                                               ##
##                                                                               ##
##              Peripheral-NTM for MPSoC                                         ##
##              Neural Turing Machine for MPSoC                                  ##
##                                                                               ##
###################################################################################

###################################################################################
##                                                                               ##
## Copyright (c) 2022-2023 by the author(s)                                      ##
##                                                                               ##
## Permission is hereby granted, free of charge, to any person obtaining a copy  ##
## of this software and associated documentation files (the "Software"), to deal ##
## in the Software without restriction, including without limitation the rights  ##
## to use, copy, modify, merge, publish, distribute, sublicense, and/or sell     ##
## copies of the Software, and to permit persons to whom the Software is         ##
## furnished to do so, subject to the following conditions:                      ##
##                                                                               ##
## The above copyright notice and this permission notice shall be included in    ##
## all copies or substantial portions of the Software.                           ##
##                                                                               ##
## THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR    ##
## IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,      ##
## FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   ##
## AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        ##
## LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, ##
## OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     ##
## THE SOFTWARE.                                                                 ##
##                                                                               ##
## ============================================================================= ##
## Author(s):                                                                    ##
##   Paco Reina Campo <pacoreinacampo@queenfield.tech>                           ##
##                                                                               ##
###################################################################################

import numpy as np

def accelerator_tensor_inverse(data_in):

  m, n, p = data_in.shape

  matrix_in_int = np.zeros((n, p))
  tensor_in_int = np.zeros((m, n, 2*p))

  data_out = np.zeros((m, n, p))

  # Augmenting Identity Matrix of Order SIZE_IN
  for i in range(m):
    for j in range(n):
      for k in range(p):
        tensor_in_int[i][j][k] = data_in[i][j][k]

        if i == j and j == k and k == i:
          tensor_in_int[i][j][k + p] = 1
        else:
          tensor_in_int[i][j][k + p] = 0

  for i in range(m):
    # Row swapping
    u = 1

    while data_in[i][i][i] == 0:
      for j in range(n):
        for k in range(p):
          matrix_in_int[j][k] = tensor_in_int[i][j][k]

      if i < m - 1:
        for j in range(n):
          for k in range(p):
            tensor_in_int[i][j][k] = tensor_in_int[i + u][j][k]

        for j in range(n):
          for k in range(p):
            tensor_in_int[i + u][j][k] = matrix_in_int[j][k]
      else:
        for j in range(n):
          for k in range(p):
            tensor_in_int[i][j][k] = tensor_in_int[i - u][j][k]

        for j in range(n):
          for k in range(p):
            tensor_in_int[i - u][j][k] = matrix_in_int[j][k]

      u += 1

    # Applying Gauss Jordan Elimination
    for j in range(n):
      for k in range(p):
        if j != k:
          scalar_ratio_int = tensor_in_int[i][k][j] / tensor_in_int[i][j][j]

          for v in range(2*p):
            scalar_sum_int = scalar_ratio_int*tensor_in_int[i][j][v]

            tensor_in_int[i][j][v] -= scalar_sum_int

  # Row Operation to Make Principal Diagonal to 1
  for i in range(m):
    for j in range(n):
      for k in range(p, 2*p):
        tensor_in_int[i][j][k] = tensor_in_int[i][j][k] / tensor_in_int[i][i][i]

  # Output
  for i in range(m):
    for j in range(n):
      for k in range(p):
        data_out[i][j][k] = tensor_in_int[i][j][k + p]

  return data_out

File: configuration/rust/library/algebra/src/tensor/accelerator_tensor_inverse.rs
///////////////////////////////////////////////////////////////////////////////////
//                                            __ _      _     _                  //
//                                           / _(_)    | |   | |                 //
//                __ _ _   _  ___  ___ _ __ | |_ _  ___| | __| |                 //
//               / _` | | | |/ _ \/ _ \ '_ \|  _| |/ _ \ |/ _` |                 //
//              | (_| | |_| |  __/  __/ | | | | | |  __/ | (_| |                 //
//               \__, |\__,_|\___|\___|_| |_|_| |_|\___|_|\__,_|                 //
//                  | |                                                          //
//                  |_|                                                          //
//                                                                               //
//                                                                               //
//              Peripheral-NTM for MPSoC                                         //
//              Neural Turing Machine for MPSoC                                  //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////////
//                                                                               //
// Copyright (c) 2020-2024 by the author(s)                                      //
//                                                                               //
// Permission is hereby granted, free of charge, to any person obtaining a copy  //
// of this software and associated documentation files (the "Software"), to deal //
// in the Software without restriction, including without limitation the rights  //
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell     //
// copies of the Software, and to permit persons to whom the Software is         //
// furnished to do so, subject to the following conditions:                      //
//                                                                               //
// The above copyright notice and this permission notice shall be included in    //
// all copies or substantial portions of the Software.                           //
//                                                                               //
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR    //
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,      //
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   //
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        //
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, //
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     //
// THE SOFTWARE.                                                                 //
//                                                                               //
// ============================================================================= //
// Author(s):                                                                    //
//   Paco Reina Campo <pacoreinacampo@queenfield.tech>                           //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

// Tensor Inversion
pub fn accelerator_tensor_inverse(data_in: &mut Vec<Vec<Vec<f64>>>) -> Vec<Vec<Vec<f64>>>{
    let length = data_in.len();

    let mut augmented_tensor = accelerator_zero_tensor(length, length, 2*length);

    for i in 0..length {
        for j in 0.. length {
            for k in 0.. length {
                augmented_tensor[i][j][k] = data_in[i][j][k];
            }
            augmented_tensor[i][j][j + length] = 1.0;
        }
    }

    accelerator_gauss_jordan_general(&mut augmented_tensor);

    let mut data_out = accelerator_zero_tensor(length, length, length);

    for i in 0..length {
        for j in 0..length {
            for k in 0..length {
                data_out[i][j][k] = augmented_tensor[i][j][k+length];
            }
        }
    }
    data_out
}

// Generalised Reduced Row Echelon Form
pub fn accelerator_gauss_jordan_general(data_in: &mut Vec<Vec<Vec<f64>>>) {
    let mut lead = 0;

    let i_counter = data_in.len();
    let j_counter = data_in[0].len();
    let k_counter = data_in[0][0].len();

    for i in 0..i_counter {
        for r in 0..j_counter {
            if k_counter <= lead {
                break;
            }

            let mut j = r;

            while data_in[i][j][lead] == 0.0 {
                j = j + 1;

                if j_counter == j {
                    j = r;
                    lead = lead + 1;

                    if k_counter == lead {
                        break;
                    }
                }
            }

            let temporal = data_in[i][j].to_owned();

            data_in[i][j] = data_in[i][r].to_owned();
            data_in[i][r] = temporal.to_owned();

            if data_in[i][r][lead] != 0.0 {
                let divider_ratio = data_in[i][r][lead];

                for k in 0..k_counter {
                    data_in[i][r][k] = data_in[i][r][k] / divider_ratio;
                }
            }

            for m in 0..j_counter {
                if m != r {
                    let multiplier_ratio = data_in[i][m][lead];

                    for k in 0..k_counter {
                        data_in[i][m][k] = data_in[i][m][k] - data_in[i][r][k] * multiplier_ratio;
                    }
                }
            }
            lead = lead + 1;
        }
    }
}

// Zero Tensor
pub fn accelerator_zero_tensor(size_i_in: usize, size_j_in: usize, size_k_in: usize) -> Vec<Vec<Vec<f64>>> {
    let mut data_in = Vec::with_capacity(size_i_in);

    for _ in 0..size_i_in {
        let mut matrix = Vec::with_capacity(size_j_in);

        for _ in 0..size_j_in {
            let mut vector = Vec::with_capacity(size_k_in);

            for _ in 0..size_k_in {
                vector.push(0.0);
            }
            matrix.push(vector);
        }
        data_in.push(matrix);
    }
    data_in
}

// Printed Tensor
pub fn accelerator_print_tensor(data_in: &mut Vec<Vec<Vec<f64>>>) {
    for i in 0..data_in.len(){
        for j in 0..data_in.len(){
            println!("{:?}", data_in[i][j]);
        }
        println!("\n");
    }
}
